### DISCUSSION TITLE
Filesystem music provider

### DISCUSSION THREAD

**Discussion by OzGav** [music-assistant/support#820] [Music providers]
Music Assistant has full support for reading your own (local) music files on disk or a remote server

Documentation https://music-assistant.io/music-providers/filesystem/

**Comment by rvandet** [comment]
I would like to use SMBv1 because i have an old PopcornHour device at home. Can i tweak it to be able to?

**Reply by marcelveldt** [reply]
The docs were outdated, smb1 is supported

**Reply by rvandet** [reply]
in the variables I added: ",vers=1.0"
and that did the trick.

**Reply by marcelveldt** [reply]
I was about to answer you that you should add vers=1.0 to the advanced options but you found out already, great!

**Comment by rvandet** [comment]
I have a RaspPi with Moode audio running Airplay, DLNA and Spotify connect.
Would there be a preferred scenario regarding streaming Lossless to Moode?

**Reply by marcelveldt** [reply]
try dlna first if that fails, use airplay (airplay is your safe bet but limited to CD quality)

**Reply by rvandet** [reply]
And how do you rank UPNP?

**Reply by OzGav** [reply]
UPNP is not a media streaming protocol https://www.linksys.com/ca/support-article/?articleNum=138198

**Comment by rvandet** [comment]
Which way is prefered use case for Spotify?
- Use in Music assistant,
- Use in Home assistant?

**Reply by OzGav** [reply]
I think that is up to you. It depends on what you want to be able to do and whether one of the platforms allows you to do that.

**Comment by antoinevandenhurk** [comment]
How do I config a local media path in Music Assistant V2, when a mounted NAS is integrated?

**Reply by OzGav** [reply]
The file system mounting is done by MA so you just add the necessary details to the file system provider. Any other mount point you have can be removed.

**Comment by rvandet** [comment]
From browsing the internet, i read that it is impossible in HA to play a folder. Are you planning to make this avaible in MA? Or did i miss the option.

**Reply by OzGav** [reply]
The typical expected setup of folders is that they would contain an album or an artist and MA is able to be used to search and play all music of that album or artist and so on that basis you can effectively play a folder. Having said that you aren’t the first to request this so if you can identify a clear use case then you are free to create a feature request and we can see how many upvotes it gets.

**Reply by rvandet** [reply]
Done so: https://github.com/orgs/music-assistant/discussions/1256

**Comment by rudi159** [comment]
I got an error "**Connection failed for the given details: SMB mount failed with error: /bin/sh: 1: Syntax error: Unterminated quoted string**"
when I was adding a music provider "Filesystem (remote share)"

Cause: password for accessing the share contained a " (quote)
Solution: change the password

**Comment by rvandet** [comment]
I want to connect with SAMBA to my TPLINK (secondary) router which has a USB connected HDD with all my music on it. MA says: "Cant resolve host". What would you advise me. In TPLINK it says "Samba for Windows", would that be the issue here?

EDIT:
Found it, had to set the IP adress only, not preceding with the two slashes.

**Comment by gon0** [comment]
Hello, maybe off-topic: I am not sure, if this is a good place to describe the issue: I have mp3 files, e.g. which I downloaded from Youtube. I added custom ID3 tags for Title, Album and Artist Name.

Files, that show the normal screen 
![grafik](https://github.com/music-assistant/hass-music-assistant/assets/33236024/c5fdbe82-e8d9-44ea-95c5-e580c0094886)
can also be found via music assistant's search.

I do not find my "custom files" with custom ID3-Tag via the Music Assistant search. If I look these files up via the folder browser, a blank screen is shown: 
![grafik](https://github.com/music-assistant/hass-music-assistant/assets/33236024/e4f2de87-f9c8-49cb-9c21-c67436d256b6)

How can I access these files? The file path to these files is shorter than 100 characters (just in case this has an effect).

**Reply by OzGav** [reply]
Look in the logs for errors. You probably don’t have the tagged properly, missing album artist tag is likely.

**Comment by hundredfire** [comment]
Hello. I have a noob question, sorry if it is dumb. I managed to connect a network share with multiple folders which contain my music. I want to shuffle a single folder in an automation, but can't figure out what to fill in the Media ID field.
The media folder that I want to play is identified as "filesystem_smb--5fJTuHbo://Music/Genshin 2nd selection/" in Music assistant's browser (and I can manually play the tracks if I select them from Music assistant's UI.
When I try to call a Play media (advanced) action, I added this url in the media ID field, but it fails with 'Failed to perform the action mass.play_media. No playable items found'
What am I doing wrong?

**Reply by OzGav** [reply]
MA does not revolve around the concept of folders, items need to be added to the library. You can create a playlist that includes all the desired tracks and then that can be added to a queue with shuffle enabled.

**Reply by hundredfire** [reply]
Thank you, 

> MA does not revolve around the concept of folders, items need to be added to the library. You can create a playlist that includes all the desired tracks and then that can be added to a queue with shuffle enabled.

Thanks, that's what I thought. However, when I tried adding a track to the library, it gave me this error: `Field "provider_mappings" of type set[ProviderMapping] is missing in PlaylistTrack instance`
Should I file a bug report or did I miss something?

**Reply by hundredfire** [reply]
Here's the relevant part of the log
[music-assistant.log](https://github.com/user-attachments/files/17808609/music-assistant.log)

**Reply by OzGav** [reply]
> when I tried adding a track to the library

What are you doing though? Tracks are added automatically when MA scans the folder.

**Reply by hundredfire** [reply]
> > when I tried adding a track to the library
> 
> What are you doing though? Tracks are added automatically when MA scans the folder.

omg I'm so dumb. It's indeed already in the library. Thank you and sorry for wasting your time. Cheers!

**Comment by Teddydm71** [comment]
HI, 
two weeks ago I installed music assistant server on Home Assistant. I activated Filesystem (local disk) as Music Provider and added my local folder. Everything seemed ok to me, then I realized that some songs were associated with the wrong artists. When in doubt I checked the tags but they were correct. I cleared the cache, deleted the database and reloaded the music folder. I tried and tried and for some unknown reason, all the tag data is correct except the artist. I've seen that some albums/tracks are associated with another artist, which sometimes changes. I'm going crazy :-) However, if I only load that singer's folder the data is correct. I also tried removing and installing Music Assistant Server. Nothing.

![Immagine 2025-02-11 214044](https://github.com/user-attachments/assets/38302bd9-904d-432f-8520-128c3b87267b)

![Immagine 2025-02-11 214204](https://github.com/user-attachments/assets/11a90498-6e05-4479-ac4f-a60cc74e4b1f)

**Reply by OzGav** [reply]
There is a bug with caused by tracks that have multiple artists. Do you have any tracks which have both Claudio and Alex?

**Reply by Teddydm71** [reply]
Hi OzGav, thanks for your quick response. There is only one track in the multiple artist folder and it is correct. The strange thing is that if I delete the db and reload the music folder instead of alex britti I find another artist. I can check if other tracks with multiple artists have the same problems.

![Immagine 2025-02-12 070403](https://github.com/user-attachments/assets/e9323dde-67d1-4232-8e5f-529e3d346a97)

**Reply by OzGav** [reply]
Its not related to a multiple artists folder. It happens with any track where there are two artists tagged. Also I am assuming you have tagged these tracks with Picard or some other software which includes all of the MusicBrainz tags as per our docs?

**Reply by Teddydm71** [reply]
Yes exactly. I used Picard

**Reply by OzGav** [reply]
Then keep investigating the tracks. You can do a minimal install as you did when it worked then add a folder at a time until something goes wrong then look at those tracks. Having said that it might be dependent on the order that the tracks are parsed.

**Reply by Teddydm71** [reply]
OK, thanks a lot. I'll try it this way. To check what has actually been loaded into the db, is there a way to directly access the folder it is in? Thanks again

**Reply by OzGav** [reply]
We don't explain to people how to access the database as that info will then be used by people who don't know what they are doing to damage their DB and then ask for help. So if you know how to access docker containers on HA and you know how to inspect databases then you will be able to work it out.

Better that you just do some testing and narrow down the problem, post a good reproducible descrption and the devs will sort it out from there.

**Reply by Teddydm71** [reply]
> We don't explain to people how to access the database as that info will then be used by people who don't know what they are doing to damage their DB and then ask for help. So if you know how to access docker containers on HA and you know how to inspect databases then you will be able to work it out.

wonderful :-) 👍 

> Better that you just do some testing and narrow down the problem, post a good reproducible descrption and the devs will sort it out from there.

Okay, I'll try to figure out the problem in the next few days. Thanks for everything.

**Reply by Teddydm71** [reply]
Here I am :-)
This evening I did some testing using a song with two artists. If I take the data as it is downloaded from Picard this is the default view:
**Artist Tag** with union of all artists;
**Albumartist Tag** with Various Artists (in this case);
**Artists Tag** with artists name separeted by semi colon;

![Picard Default](https://github.com/user-attachments/assets/055dae6f-1002-422f-a65b-fe2d72da0617)

Then if i click on Artists Tag the app open this windows
![Picard Default - List](https://github.com/user-attachments/assets/927b3c57-14ad-4919-ba00-fa287207665a)

but if i open the file the artists name are separeted by null char
![Picard Default - List - Null](https://github.com/user-attachments/assets/52d9c8a3-ff01-45d3-834e-8976e43f864b)

when i load the mp3 file in the music assistant server this is the situation:
There are only Elton John and Various Artists as Artists
![Picard Default 1](https://github.com/user-attachments/assets/8b436163-02ad-4f03-9083-c8986343fc44)
the album is correct
![Picard Default 2](https://github.com/user-attachments/assets/70b20f04-d263-401e-81d4-bf6b4b4dbb10)

the track is assigned only to Elton John
![Picard Default 3](https://github.com/user-attachments/assets/99ee3ca9-0bbc-4342-b81b-7825005f9087)

If i change the Artists Tag in 
![List - New](https://github.com/user-attachments/assets/246a6145-bcd6-44f8-9813-c04b05d2d446)

in Picard i see the same situation
![List - New - View](https://github.com/user-attachments/assets/ffd00b23-fc8c-47ce-9bcc-d1e70b55ad78)

but the string is written correctly as expected by the music assistant server parser
![Picard Default - List - No Null](https://github.com/user-attachments/assets/f7e65f37-e92b-430d-b197-cae74c82c6fc)
 and so, after clear cache and delete database we have:

![Picard New](https://github.com/user-attachments/assets/d8616193-affc-4040-956e-0e221e7f0182)
![Picard New 2](https://github.com/user-attachments/assets/6f77152b-dc18-47e3-a85a-8f4b897623cb)
![Picard New 3](https://github.com/user-attachments/assets/72811874-48d9-4c60-9cd3-98247d4c2409)

### ** All is correct**

therefore, either I modify all the tags of the mp3 files with more than one artist or the parser is implemented providing this method of separation between the names of the artists :-)

**Reply by OzGav** [reply]
This is the known issue I was talking about.

**Reply by marcelveldt** [reply]
This happens when you use ID3v2.4 tags which are not fully supported by every tag reader. For best compatibility, use 2.3 tags.
We have some plans to upgrade our tag reader at some point which will fix this issue.

**Reply by marcelveldt** [reply]
I actually implemented a quick workaround for the 2.4 tags (well at least, the most important ones) by manually parsing the frames. So at least issues like this will be gone when people have the version set to 2.4

**Reply by Teddydm71** [reply]
> I actually implemented a quick workaround for the 2.4 tags (well at least, the most important ones) by manually parsing the frames. So at least issues like this will be gone when people have the version set to 2.4

Perfec!!! Thank you so much.

**Reply by OzGav** [reply]
2.4.0rc1 has been released if you want to test this

**Reply by Teddydm71** [reply]
> 2.4.0rc1 has been released if you want to test this

Hi Ozgav, i have downloaded and installed 2.4.0.rc2, it's the same? 
I tried with 2 Michael Jackson songs. Picard tags set to 2.4.
MA only uploaded one song... the version with Justin Timberlake was discarded.

![tag](https://github.com/user-attachments/assets/3e0b6787-5cce-4ee6-8108-c7a419b6b3ef)
![duet](https://github.com/user-attachments/assets/1cc16ea1-8c31-49e6-91a1-294e72f2ec3d)
![original](https://github.com/user-attachments/assets/6dd5d33c-69dc-4c39-9845-212360680d6d)

**Reply by OzGav** [reply]
Do you have MusicBrainz tags added to those tracks? Your image doesn't show them

**Reply by Teddydm71** [reply]
> Do you have MusicBrainz tags added to those tracks? Your image doesn't show them

Yes, sorry, I used the MusicBrainz Picard tags without any changes compared to the original

**Reply by OzGav** [reply]
Please show all of the tags

**Reply by Teddydm71** [reply]
> Please show all of the tags

![Tag](https://github.com/user-attachments/assets/816c0314-430c-4e16-81ef-267929ea5504)

These are the original tags of MusicBrainz Picard.
My tag label are in italian, but the the underline original tag is %artists%.

**Reply by OzGav** [reply]
Are you sure this hasn't appeared under a different name such as Justin Timberlake or Michael Jackson feat. Justin Timberlake.? What do you see in the logs?

**Reply by Teddydm71** [reply]
Yes, on MA beta i have loaded only MJ Folder...
I have only 2 artists in my MA Server.
For your information i have stopped MA Server 2.3.6 and i run MA Server 2.4.0 rc2

![Immagine 2025-02-18 130637](https://github.com/user-attachments/assets/6e4d59f7-45ea-464d-9c1a-32be3ab122bd)

**Reply by OzGav** [reply]
What about the logs? You can try removing the album from the library so it will get reimported and see what is reported in the log.

**Reply by Teddydm71** [reply]
> What about the logs? You can try removing the album from the library so it will get reimported and see what is reported in the log.

![error log](https://github.com/user-attachments/assets/306bf4a0-55bd-4593-a128-693f5782b05b)

**Reply by OzGav** [reply]
Yeah I would say the 2.4 tags are still a problem. Try the next RC when it comes out.

**Reply by Teddydm71** [reply]
> Yeah I would say the 2.4 tags are still a problem. Try the next RC when it comes out.
Thank you. :-)

**Reply by Teddydm71** [reply]
All perfect!!! With MA rc4 the 2.4 tags are full compliance. Thank you so much

**Comment by Hardy3eich** [comment]
how to play music from attaches USB-Stick?
I do not want to take the long way as network attached device as the stick is direktly attachet to my unraid server.
my installation is docker in unraid.
thank you for help

**Reply by OzGav** [reply]
You need th path to the files https://www.music-assistant.io/music-providers/filesystem/

**Reply by Hardy3eich** [reply]
ok got this "Docker users can mount their own folder paths"
but how do I do that exactly within unraid? 
usb-stick is mounted in unraid as
Identification: Generic_Flash_Disk_22050470050308-0:0 (sdd) 
mount point: mnt/disks/NTFS256gb

**Reply by OzGav** [reply]
You need to ask this on a dedciated unraid forum. We just deal with MA queries.

**Reply by Hardy3eich** [reply]
ok no problem but am I correct:
MA filesystem (local disk) can only access to folder
/media

**Reply by OzGav** [reply]
No as far as I know you can put what ever path in there. Maybe you could set up a symlink or something?

**Comment by UKenGB** [comment]
The Local filesystem provider has 'Reload' and 'Synchronise now' options. What's the difference?

I couldn't find any mention of this in the docs, but the terms rather imply that the former throws away the data and starts scanning all over again while the latter just updates the Library with changes in the filesystem. Is that the case or is my understanding incorrect?

**Reply by OzGav** [reply]
Reload shuts down the provider and restarts it. Synchronise Now just starts the sync task

**Reply by UKenGB** [reply]
👍

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
